Training/python/2018v1/Training_v0p1.py

```diff
@@ -43,64 +46,131 @@ def __init__(self, *args, loss=None, **kwargs):
-        self.loss_tracker.update_state(loss, sample_weight=sample_weight)
-        self.pure_loss_tracker.update_state(pure_loss, sample_weight=sample_weight) 
-        self.reg_loss_tracker.update_state(reg_loss)
-        self.compiled_metrics.update_state(y, y_pred, sample_weight)
+        # Forward Pass:
+        if self.use_AdvDataset:
+            with tf.GradientTape() as class_tape, tf.GradientTape() as adv_tape:
+                    y_pred = self(x, training=True)
+                    reg_losses = self.losses 
+                    pure_loss_vec = self.model_loss(y, y_pred[0])
+                    pure_loss = tf.reduce_sum(tf.multiply(pure_loss_vec, sample_weight))/self.n_tau
+                    adv_loss_vec = self.adv_loss(y_adv, y_pred[1])
+                    adv_loss = tf.reduce_sum(tf.multiply(adv_loss_vec, sample_weight_adv[:,0]))/self.n_tau
+                    if reg_losses:
+                        reg_loss = tf.add_n(reg_losses)
+                        loss = pure_loss + reg_loss
+                        loss_vec = pure_loss_vec + reg_loss
+                    else:
+                        reg_loss = reg_losses # Empty
+                        loss = pure_loss
+                        loss_vec = pure_loss_vec
+        else:
+            with tf.GradientTape() as class_tape:
+                y_pred = self(x, training=True)
+                reg_losses = self.losses
+                pure_loss_vec = self.model_loss(y, y_pred)
+                pure_loss = tf.reduce_sum(tf.multiply(pure_loss_vec, sample_weight))/self.n_tau
+                if reg_losses:
+                    reg_loss = tf.add_n(reg_losses)
+                    loss = pure_loss + reg_loss
+                    loss_vec = pure_loss_vec + reg_loss
+                else:
+                    reg_loss = reg_losses # Empty
+                    loss = pure_loss
+                    loss_vec = pure_loss_vec
+        # Compute gradients and update weights
+        if self.use_AdvDataset:
+            class_layers = [var for var in self.trainable_variables if ("final" in var.name and "_adv" not in var.name)] # final classification dense only
+            adv_layers = [var for var in self.trainable_variables if ("final" in var.name and "_adv" in var.name)] #final adv only
+            common_layers = [var for var in self.trainable_variables if "final" not in var.name] #gradients common to both
+            grad_class = class_tape.gradient(loss, common_layers + class_layers) 
+            grad_adv = adv_tape.gradient(adv_loss, common_layers + adv_layers)
+            grad_class_excl = grad_class[len(common_layers):] # gradients of common part
+            grad_adv_excl = grad_adv[len(common_layers):] #gradients of adv part
+            grad_common = [self.k1*grad_class[i] - self.k2 * grad_adv[i] for i in range(len(common_layers))] 
+            mean_common = [tf.math.reduce_mean(tf.math.abs(grad_common[i])) for i in range(len(grad_common))] 
```

Comment:
- should be split into two metrics `mean_grad_class` and `mean_grad_common`
- I propose to add `tf.add_n(..) / len(grad_common)` to have a scalar value

Reply:
Assuming you meant mean_grad_class and mean_grad_adv, done